flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPoolTest.java

```diff
@@ -242,23 +243,129 @@ public void testRecycleAfterDestroy() {
+        bufferPool.setNumBuffers(4);
+        assertThat(bufferPool.getNumBuffers()).isEqualTo(4);
+        assertThat(bufferPool.getNumberOfAvailableMemorySegments()).isZero();
+        assertThat(bufferPool.getNumberOfRequestedOverdraftMemorySegments()).isOne();
+        assertThat(bufferPool.isAvailable()).isFalse();
+        buffers.add(bufferPool.requestMemorySegmentBlocking());
+        assertThat(bufferPool.getNumberOfRequestedOverdraftMemorySegments()).isEqualTo(2);
+        assertThat(bufferPool.isAvailable()).isFalse();
+
+        // return all overdraft buffers.
+        bufferPool.recycle(buffers.poll());
+        assertThat(bufferPool.getNumberOfRequestedOverdraftMemorySegments()).isOne();
+        assertThat(bufferPool.isAvailable()).isFalse();
+        bufferPool.recycle(buffers.poll());
+        assertThat(bufferPool.getNumberOfRequestedOverdraftMemorySegments()).isZero();
+        assertThat(bufferPool.isAvailable()).isFalse();
+
+        // return the excess buffer.
+        bufferPool.recycle(buffers.poll());
+        assertThat(bufferPool.isAvailable()).isFalse();
+        // return non-excess buffers.
+        bufferPool.recycle(buffers.poll());
+        assertThat(bufferPool.getNumberOfAvailableMemorySegments()).isOne();
+        assertThat(bufferPool.isAvailable()).isTrue();
+
+        while (!buffers.isEmpty()) {
+            bufferPool.recycle(buffers.poll());
+        }
+        bufferPool.lazyDestroy();
+    }
+
+    @Test
+    void testIncreasePoolSize() throws Exception {
+        LocalBufferPool bufferPool =
+                new LocalBufferPool(networkBufferPool, 5, 100, 0, Integer.MAX_VALUE, 2);
+        List<MemorySegment> buffers = new ArrayList<>();
+
+        // set pool size to 5.
+        bufferPool.setNumBuffers(5);
+        assertThat(bufferPool.getNumBuffers()).isEqualTo(5);
+
+        // request all buffer.
+        for (int i = 0; i < 5; i++) {
+            buffers.add(bufferPool.requestMemorySegmentBlocking());
+        }
+        assertThat(bufferPool.isAvailable()).isFalse();
+
+        // request 2 overdraft buffers.
+        buffers.add(bufferPool.requestMemorySegmentBlocking());
+        buffers.add(bufferPool.requestMemorySegmentBlocking());
+        assertThat(bufferPool.requestMemorySegment()).isNull();
+        assertThat(bufferPool.getNumberOfRequestedOverdraftMemorySegments()).isEqualTo(2);
+        assertThat(bufferPool.isAvailable()).isFalse();
+
+        // set pool size to 10.
+        bufferPool.setNumBuffers(10);
+        assertThat(bufferPool.getNumBuffers()).isEqualTo(10);
+        assertThat(bufferPool.getNumberOfAvailableMemorySegments()).isOne();
+        assertThat(bufferPool.getNumberOfRequestedOverdraftMemorySegments()).isEqualTo(2);
+        // available status will not be influenced by overdraft.
+        assertThat(bufferPool.isAvailable()).isTrue();
+        buffers.add(bufferPool.requestMemorySegmentBlocking());
+        assertThat(bufferPool.isAvailable()).isTrue();
```

Comment:
There may be bug here.

I change the 10 to 7, it should be unavailable, however, it's available. 

`bufferPool.setNumBuffers(7);` requested new buffer from global buffer pool to the `availableMemorySegments`.

Call stack is as follow :

```
	  at org.apache.flink.runtime.io.network.buffer.LocalBufferPool.requestMemorySegmentFromGlobal(LocalBufferPool.java:442)
	  at org.apache.flink.runtime.io.network.buffer.LocalBufferPool.checkAvailability(LocalBufferPool.java:551)
	  at org.apache.flink.runtime.io.network.buffer.LocalBufferPool.checkAndUpdateAvailability(LocalBufferPool.java:524)
	  at org.apache.flink.runtime.io.network.buffer.LocalBufferPool.setNumBuffers(LocalBufferPool.java:682)
	  - locked <0xa5d> (a java.util.ArrayDeque)
	  at org.apache.flink.runtime.io.network.buffer.LocalBufferPoolTest.testIncreasePoolSize(LocalBufferPoolTest.java:331)
```

I think `isRequestedSizeReached` should be updated.


```
    private boolean isRequestedSizeReached() {
        return numberOfRequestedMemorySegments + numberOfRequestedOverdraftMemorySegments
                >= currentPoolSize;
    }
```

WDYT?

BYW, we should add more tests here, such as : newPoolSize=7.

Reply:
Or, should we reset the `numberOfRequestedMemorySegments` and `numberOfRequestedOverdraftMemorySegments` inside of the `setNumBuffers`?

If `currentPoolSize` is increased, we should decrease the `numberOfRequestedOverdraftMemorySegments` and increase `numberOfRequestedMemorySegments`.

I'm not sure which solution is better. And I prefer reset them inside of  the `setNumBuffers`, because it's more clear, and other solutions may have bug when `returnMemorySegment`.

Reply:
Good catch !!! I think the second one is more better. If pool size becomes larger, some overdraft buffers need convert to requested buffers.